collector/agentPoller.go

```diff
@@ -15,451 +15,433 @@
+		return a, fmt.Errorf("Must pass ipAddress to NewAgent()")
 	}
 	if port < 1024 {
-		return a, errors.New("Must pass port to NewAgent()")
+		return a, fmt.Errorf("Must pass port > 1024 to NewAgent()")
 	}
 	if pollPeriod == 0 {
-		return a, errors.New("Must pass pollPeriod to NewAgent()")
-	}
-	if len(topic) == 0 {
-		return a, errors.New("Must pass topic to NewAgent()")
+		return a, fmt.Errorf("Must pass pollPeriod to NewAgent()")
 	}
 
 	a.IPCommand = ipCommand
 	a.Port = port
 	a.PollPeriod = pollPeriod
-	a.Topic = topic
-	a.AgentStateChan = make(chan *AgentState)
+	a.MetricsChan = metricsChan
+
+	// Detect the agent's IP address once. Per the DC/OS docs (at least as of
+	// November 2016), changing a node's IP address is not supported.
+	//
+	if a.AgentIP == "" { // allows us to mock this during testing
+		if a.AgentIP, err = a.getIP(); err != nil {
+			log.Error(err)
+		}
+	}
 
 	return a, nil
 }
 
-// Run runs an Agent Poller which periodically produces data retrieved from a local Mesos Agent.
-// This function should be run as a gofunc.
-func (a *Agent) Run(recordsChan chan<- *AvroDatum, stats chan<- statsd.StatsEvent) {
-	// fetch agent ip once. per DC/OS docs, changing a node IP is unsupported
-	if len(*agentTestStateFileFlag) == 0 ||
-		len(*agentTestSystemFileFlag) == 0 ||
-		len(*agentTestContainersFileFlag) == 0 {
-		// only get the ip if actually needed
-		//TODO needs err
-		a.getIP(stats)
-	}
+// RunPoller periodiclly polls the HTTP APIs of a Mesos agent. This function
+// should be run in its own goroutine.
+func (a *Agent) RunPoller() {
+	ticker := time.NewTicker(a.PollPeriod)
 
-	// do one poll immediately upon starting, to ensure that agent metadata is populated early:
-	a.pollAgent(recordsChan, stats)
-	ticker := time.NewTicker(time.Second * time.Duration(a.PollPeriod))
+	// Poll once immediately
+	for _, m := range a.transform(a.pollAgent()) {
+		a.MetricsChan <- m
+	}
 	for {
 		select {
 		case _ = <-ticker.C:
-			a.pollAgent(recordsChan, stats)
+			for _, m := range a.transform(a.pollAgent()) {
+				a.MetricsChan <- m
+			}
 		}
 	}
 }
 
-// ---
-
-func (a *Agent) pollAgent(recordsChan chan<- *AvroDatum, stats chan<- statsd.StatsEvent) {
-	// always fetch/emit agent state first: downstream will use it for tagging metrics
-	agentState, err := a.getAgentState(stats)
-	if err == nil {
-		a.AgentStateChan <- agentState
-	} else {
-		log.Printf("Failed to retrieve state from agent at %s: %s", a.AgentIP, err)
+// getContainerMetrics queries an agent for container-level metrics, such as
+// CPU, memory, disk, and network usage.
+func (a *Agent) getContainerMetrics() ([]agentContainer, error) {
+	var containers []agentContainer
+
+	// TODO(roger): 15sec timeout is a guess. Is there a better way to do this?
+	c := NewHTTPClient(
+		strings.Join([]string{a.AgentIP, strconv.Itoa(a.Port)}, ":"),
+		"/containers",
+		time.Duration(15*time.Second))
+	if err := c.Fetch(&containers); err != nil {
+		log.Error(err)
+		return nil, err
```

Comment:
I'm a big fan of returning the err right away and having the the top most caller log the err.Error()